Support for Shotcut 22

MaZderMind opened the pull request:
This PR improves Support for Shotcut 22 by implementing the new ML-XML Format which uses `<chain>` instead of `<producer>`. Also Write-Access to the Tempfiles is rejected for both the Shotcut 19+ as well as the Shotcut 22 Tempfile Naming-Schema in order to prevent QtSaveFile from performing an atomic write, which fuse-ts does not support.

Fixes #13

a-tze committed:
Change access bits for shotcut tempfiles to readonly

this prevents QtSaveFile from performing atomic writes with tempfiles which fuse-ts does not support

Commit:
Add support for Shotcut 22 style mlt-xml

Commit:
Add support for Shotcut 22 style tempfiles

Commit:
Cleanup comments

MaZderMind commented:
These changes work for Shotcut 22, however they seem to break support for Shotcut 19. Should I try to incorporate the old Schema of handling Shotcut 19 style Tempfiles or would you rather like to drop support for Shotcut 19 @a-tze ?

a-tze commented on fuse-ts-knowledge.c:
```diff
@@ -40,7 +40,10 @@ int get_index_from_pathname(const char* path) {
+	} else if (shotcut_tmp_path && (
+		(strncmp (path, "/shotcut-", 9) == 0) ||
+		(strncmp (path, "/project_shotcut.mlt.", 21) == 0)
+	)) {
```
are there usages of INDEX_SHOTCUT_TMP left in fuse-ts.c? I'm wondering if this branch cannot be removed completely now.

MaZderMind replied:
Yes there are still 3 uses left, which are used to differentiate between EACCES and ENOENT.

a-tze commented:
@MaZderMind I did not dig into the past of the shotcut code, but my expectation was that the "direct write fallback" would also apply to older shotcut versions. 
I'd rather think that the XML content itself is not compatible between the versions?

MaZderMind commented:
I tried to combine both paths, but I found Shotcut 19 Support with the Tempfile-Hack to be quite unstable on MacOS. Saving results in the App hanging and calling into Fuse over and over again. This is also the case with current master.

a-tze commented:
I see.. Shotcut 19 had a very different method of writing the project file. so with unstable and hack you mean the previous versions of fuse-ts before adapting to Shotcut 22?

MaZderMind commented:
Here is my effort to combine Shotcut 19 and 22 Support: https://github.com/a-tze/fuse-ts/compare/master...MaZderMind:fuse-ts:13-shotcut-project-file-shotcut-19+22

However, this is still broken for Shotcut 19 in the same was as the current master is.
From my point of view, it would be nicer to ditch the broken Shotcut 19 Support and only support the most recent version.

MaZderMind commented:
It seems there is still some sort of Problem with Shotcut 22: After Saving, the file is received by fuse-ts and re-opening the xml also yields the correct cutmarks, however inframe and outframe are not updated until the file is re-opened again.

a-tze commented:
@MaZderMind Agree. The old versions of fuse-ts are still available, its no force to the users and there is no other change bundled to this. Additionally the Tempfile hack might be dropped for Kdenlive as well in a later version.

You know that I will likely squash the PR?

a-tze commented:
> It seems there is still some sort of Problem with Shotcut 22: After Saving, the file is received by fuse-ts and re-opening the xml also yields the correct cutmarks, however inframe and outframe are not updated until the file is re-opened again.

Must be a remain of the tempfile hack, where the cutmarks where extracted in the rename function.

MaZderMind commented:
`find_cutmarks_in_shotcut_project_file` is called in `ts_release` which seems resonable.

I think it might be a timing problem:
when saving and closing the project, i see
```
filebuffer__write:  writing 1204 bytes from 0x7f3b61976060 to 0x7f3b54002760 with offset 0
filebuffer__truncate:  truncating filebuffer 0x7f3b54002760 to size 1204
```
and only ~40s later I get the corresponding
```
release called on '/project_shotcut.mlt'
```

which then results in the cutmarks being updated. I guess that's how it always was on MacOS but in production use nobody noticed. I don't know if it would be save to parse the file in `truncate` or not.

I don't care if you'll sqush or not – commits are part of the delivery and should be nice and coherent, also to allow for better cherry-picking or review.

a-tze commented:
I think that truncate is an internal truncate method, not the filesystem truncate syscall. So it would make sense, to try to read the cutmarks in ts_write (after truncating) as well, but ignore errors because it might have been a partial write of the file. In reality it's pretty unlikely that the project file is not written with one single call, because it is so small but there still is a chance. When release is called later, they can safely be evaluated again, as the outcome would be the same.

MaZderMind commented:
There's another Problem when the production is setup with fps != 25 – the Shotcut-File does not mention the Framerate and Shotcut defaults to 25 fps, which when the file is opened again will shift the cutmarks. The fps is known so I would just add them back again to the profile (they were removed in 0587dd364dfbb9d4e3afadf6dba72ef73d97e76e)

Commit:
Denote fps in Shotcut-File

defaults to 25, but fuse-ts supports productions with different fps.

Commit:
Try to parse shotcut-file on every write to avoid race condition with the finalize-worker

at least on MacOS the release can be delayed by up to 60s after the final write

MaZderMind commented:
I'm happy now. Let me know if you want anything to be changed :)

a-tze commented on fuse-ts-shotcut.c:
```diff
@@ -140,12 +135,16 @@ int find_cutmarks_in_shotcut_project_file (int *inframe, int *outframe, int *bla
-		debug_printf ("find_cutmarks: node with id 'producer0' not found!\n");
-		mxmlRelease (xmldoc);
-		return 2;
+		node = mxmlFindElement (xmldoc, xmldoc, "chain", "id", "chain0", MXML_DESCEND);
```
If support for Shotcut 19 is dropped, is the lookup of "producer0" still necessary or does shotcut always write a "chain0" ?

MaZderMind replied:
Support for Shotcut <22 is not generally dropped; the workflow "copy to computer, edit there, copy back" is still working and has, anecdotally, been used for quite a while. Also, we're still generating`producer0` in the template ourselfs.

If you'd like to make a clean cut, I can rearrange the template and parsing-code to only support the chain element. I did not test Shotcut 22 for Windows or Linux yet, but on MacOS Shotcut 22 always generates the chain Element in our usecase and converts producer to chain.

a-tze replied:
Yes, if basically dropping support for older versions, I'd prefer to include a more recent template to be more future-compatible. Maybe the change is not so huge and it can still be read successfully from older shotcut versions too.

It just seems like a very good opportunity to create a blank project in shotcut, drop a single file in, save that project file and strip it down to the minimum.

MaZderMind replied:
Alright, I agree. I'll update the Template tomorrow and also verify what Shotcut 19 will do with it.

MaZderMind replied:
@kunsi argued, that debian stable, which is widely used at C3VOC, still has 21.01.29, so that should be the baseline. I'll verify compatibility with 21.01.29, too.

MaZderMind replied:
Shotcut 21.01 will work fine with the Tempfile-Readonly-Solution but still saves as the `producer0`style Elements and can't open a File with `chain0` style Elements. I would therefore upgrade the Template to what Shotcut 21 expects and support 21+ without bigger changes.

MaZderMind replied:
Only difference for Shotcut 21 was the specified mlt version, which I upgraded. The amount of "legacy" code is minute (only the lookup of the `producer0` element – even the parsing code is the same and shared, so that I don't think we should drop Shotcut 21 support just for the sake of removing ~6 lines.

The latest Shotcut 21 was released end of Dec 2021; Shotcut 22 is officially released since Jan 2022.

Commit:
increment mlt format to Shotcut 21 baseline

MaZderMind commented:
Anything holding this up?

a-tze commented:
Sorry, I remembered "something still needs to be done".

So let me summarize: the tip of the PR branch supports Shotcut 21 and 22 and reintroduced fps?

MaZderMind commented:
Shotcut 22 & 21 are fully supported, lower Versions will struggle with the Temp-File Problem but can still be used via Save-As-Local-File-and-copy-over Workaround. The fps have been reintroduced for Shotcut to generate the correct Timestamps on != 25 fps Projects (we produce with 30 fps and tested with that material).

MaZderMind commented:
Ping :)

MaZderMind commented:
FTR I triggered the Build-Job in the VOC Jenkins and now up-to-date Packets are available in https://pkg.c3voc.de/pool/main/f/fuse-ts/
